File: storage/db-core/src/main/java/com/coffee/gu/event/EventLogEntity.java
package com.coffee.gu.event;

import com.coffee.gu.EventLog;
import com.coffee.gu.enums.EventLogStatus;
import com.coffee.gu.enums.EventLogTarget;
import com.coffee.gu.enums.EventType;
import com.coffee.gu.BaseCustomIdEntity;
import jakarta.persistence.*;

import java.time.LocalDateTime;

@Table(name = "event_log")
@Entity
public class EventLogEntity extends BaseCustomIdEntity<String> {

    @Enumerated(EnumType.STRING)
    private EventType eventType;
    @Enumerated(EnumType.STRING)
    private EventLogTarget eventLogTarget;
    private String payload;
    private Boolean isPublished;
    
    @Enumerated(EnumType.STRING)
    private EventLogStatus status;
    private Integer retryCount;
    
    private LocalDateTime createdAt;
    private LocalDateTime publishedAt;

    protected EventLogEntity() {
        super();
    }

    private EventLogEntity(String eventId, EventType eventType, String payload, Boolean isPublished, EventLogStatus status, Integer retryCount, LocalDateTime createdAt, LocalDateTime publishedAt, boolean isNewEntity) {
        super(eventId, isNewEntity);
        this.eventType = eventType;
        this.payload = payload;
        this.isPublished = isPublished;
        this.status = status;
        this.retryCount = retryCount;
        this.createdAt = createdAt;
        this.publishedAt = publishedAt;
    }

    public static EventLogEntity create(String eventId, EventType eventType, String payload) {
        return new EventLogEntity(
                eventId,
                eventType,
                payload,
                false,
                EventLogStatus.PENDING,
                0,
                LocalDateTime.now(),
                null,
                true);
    }

    public EventLog toModel() {
        return new EventLog(
                this.id,
                this.eventType,
                this.eventLogTarget,
                this.payload,
                this.isPublished,
                this.status,
                this.retryCount,
                this.createdAt,
                this.publishedAt);
    }

    public EventType getEventType() {
        return eventType;
    }

    public String getPayload() {
        return payload;
    }

    public Boolean getPublished() {
        return isPublished;
    }

    public EventLogStatus getStatus() {
        return status;
    }

    public Integer getRetryCount() {
        return retryCount;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getPublishedAt() {
        return publishedAt;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/event/EventLogEntity.kt
package com.coffee.gu.event

import com.coffee.gu.BaseCustomIdEntity
import com.coffee.gu.EventLog
import com.coffee.gu.enums.EventLogStatus
import com.coffee.gu.enums.EventLogTarget
import com.coffee.gu.enums.EventType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.LocalDateTime

@Table(name = "event_log")
@Entity
class EventLogEntity(
    @Id
    val eventId: String,
    @Enumerated(EnumType.STRING)
    val eventType: EventType,
    @Enumerated(EnumType.STRING)
    val eventLogTarget: EventLogTarget = EventLogTarget.INTERNAL,
    val payload: String,
    var isPublished: Boolean = false,
    @Enumerated(EnumType.STRING)
    var status: EventLogStatus = EventLogStatus.PENDING,
    var retryCount: Int = 0,
    var publishedAt: LocalDateTime? = null,
    isNewEntity: Boolean = true,
) : BaseCustomIdEntity<String>(isNewEntity) {

    override fun getId(): String = eventId

    fun toModel(): EventLog {
        return EventLog(
            eventId = eventId,
            eventType = eventType,
            eventLogTarget = eventLogTarget,
            payload = payload,
            isPublished = isPublished,
            status = status,
            retryCount = retryCount,
            createdAt = createdAt,
            publishedAt = publishedAt,
        )
    }

    companion object {
        @JvmStatic
        fun create(
            eventId: String,
            eventType: EventType,
            payload: String,
            eventLogTarget: EventLogTarget = EventLogTarget.INTERNAL,
        ): EventLogEntity {
            return EventLogEntity(
                eventId = eventId,
                eventType = eventType,
                eventLogTarget = eventLogTarget,
                payload = payload,
                isPublished = false,
                status = EventLogStatus.PENDING,
                retryCount = 0,
                publishedAt = null,
                isNewEntity = true,
            )
        }
    }
}
